File: BEmu_cpp/src/MarketDataRequest/MarketElementDatetime.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/MarketDataRequest/MarketElementDatetime.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "MarketDataRequest/MarketElementDatetime.h"
#include "Types/DisplayFormats.h"

namespace BEmu
{
	namespace MarketDataRequest
	{
		MarketElementDatetime::MarketElementDatetime(const std::string& name, const Datetime& datetime)
		{
			this->_name = name;
			this->_datetime = datetime;
		}

		Name MarketElementDatetime::name() const
		{
			return Name(this->_name.c_str());
		}

		size_t MarketElementDatetime::numValues() const
		{
			return 1;
		}

		size_t MarketElementDatetime::numElements() const
		{
			return 0;
		}

		bool MarketElementDatetime::isNull() const
		{
			return false;
		}

		Datetime MarketElementDatetime::getValueAsDatetime(int index) const
		{
			if(index == 0)
				return this->_datetime;
			else
				throw elementPtrEx;
		}

		std::ostream& MarketElementDatetime::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string strDateTime(DisplayFormats::MarketDataRequests_FormatDateZone(this->_datetime));
			this->prettyPrintHelper(stream, level, spacesPerLevel, strDateTime);
			return stream;
		}
		
	}
}

File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/MarketElementDatetime.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/MarketElementDatetime.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.DateTimeTypeEnum;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Schema;

public class MarketElementDatetime extends Element
{
	private final Datetime _date;
	private final String _name;
	
	MarketElementDatetime(String name, Datetime date)
    {
		this._date = new Datetime(date, DateTimeTypeEnum.both);		
        this._name = name;
    }
	
    public Name name() throws Exception
    {
    	return new Name(this._name);
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }

    public Schema.Datatype datatype()
    {
    	return Schema.Datatype.DATETIME;
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
    	return super.prettyPrintHelper(tabIndent, this._date.toString());
    }
    
    public Object getValue()
    {
    	return this._date;
    }

    public Datetime getValueAsDatetime()
    {
        return this._date;
    }

    public Datetime getValueAsDate() //note that this does not truncate the value to date-only, it still returns both the date and the time
    {
        return this._date;
    }

    public Datetime getValueAsTime() //note that this does not truncate the value to date-only, it still returns both the date and the time
    {
        return this._date;
    }

}
